spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -501,24 +502,24 @@ else if (result instanceof Message) {
 		}
 	}
 
-	private Message<?> checkHeaders(Object result, String topic, @Nullable Object source) { // NOSONAR (complexity)
-		Message<?> reply = (Message<?>) result;
+	private Message<?> checkHeaders(Message<?> reply, @Nullable String topic, @Nullable Object source) { // NOSONAR (complexity)
 		MessageHeaders headers = reply.getHeaders();
-		boolean needsTopic = headers.get(KafkaHeaders.TOPIC) == null;
+		boolean needsTopic = topic != null && headers.get(KafkaHeaders.TOPIC) == null;
 		boolean sourceIsMessage = source instanceof Message;
-		boolean needsCorrelation = headers.get(this.correlationHeaderName) == null && sourceIsMessage;
-		boolean needsPartition = headers.get(KafkaHeaders.PARTITION) == null && sourceIsMessage
+		boolean needsCorrelation = sourceIsMessage && headers.get(this.correlationHeaderName) == null
+				&& ((Message<?>) source).getHeaders().get(this.correlationHeaderName) != null;
+		boolean needsPartition = sourceIsMessage && headers.get(KafkaHeaders.PARTITION) == null
```

Comment:
I moved `xx != null` testing to the beginning for it is much more light-weight than `headers.get(...) != null`, so shot-circuiting might avoid some unnecessary cpu cycles. Someone even claims such nullness checking involves almost zero cost in openJDK.